# Population stratification 

The single nucleotide polymorphisms (SNPs) are genetic variations which determine the diference between any two unrelated individuals, the SNPs can be used to identify the correct source population of an individual. For eficient population identification with the HapMap genotype data, as few informative SNPs as possible are required from the original 4 million SNPs. Recently, Park et al. (2006) adopted the nearest shrunken centroid method to classify the three populations, i.e., Utah residents with ancestry from Northern and Western Europe (CEU), Yoruba in Ibadan, Nigeria in West Africa (YRI), and Han Chinese in Beijing together with Japanese in Tokyo (CHB + JPT), from which 100736 SNPs were obtained and the top 82 SNPs could completely classify the three populations.

## Goal
Perform a multivariate approach of the dataset such as visualization and clustering, and to build predictive models that could classify the given populations. 

## Methodology
As classification algorithms
 
 - k-nearest neighbor (k-NN)
 
 - Linear Discriminative Analysis
 
 - Quadratic Discriminative Analysis 
 
 - Naive Bayes
 
 - Support Vector Machine (SVM)
 
 - Decision Trees 
 
 - Random Forest 
 
 were used to evaluate the success of the models according to their accuracy, specificity, sensitivity.
